### PR TITLE
Add `record where` syntax sugar

### PR DESCRIPTION
**Pull request by rhendric**
This PR implements some of the desired features in #4275; it adds a new `record where` expression which may hold bindings that would be valid in a `let`, and any names thus bound (including imported names and names inside LHS patterns) are used as fields in the record. The following

```agda
record where
  open <m> using (<using>) renaming (<x> to <y>)
  <lhs1> = <rhs1>
  <lhs2> = <rhs2>
```
desugars to

```agda
let
  <using> = <m>.<using>
  <y> = <m>.<x>
  <lhs1> = <rhs1>
  <lhs2> = <rhs2>
in record { <using> = <using> ; <y> = <y> ; <lhs1> = <lhs1> ; <lhs2> = <lhs2> }
```

This gives the following benefits:
* No type signatures required, even for functions
* Later bindings can use earlier bindings
* Extra declarations generate warnings
* Get a slightly verbose record update syntax at no additional cost:
```agda
p2 = record where
  open Pair p1 using (fst)
  snd = 0
```

It also bears some limitations:
* Everything that trips people up about the difference between `let` and `where` is relevant here. In particular, the proposed examples using `open import` don't work as written because `import` isn't valid in a `let`. `import Some.Module as M` in an outer scope and `open M using (x)` inside the record achieves the same thing. #6149 may interact with this feature in interesting (hopefully mostly positive, but who knows) ways.
* Most of what trips people up about shadowing names is relevant here, if the field accessor functions are imported in an outer scope. Note however that `open` statements are translated such that later references to names of imported fields refer unambiguously to the locally-bound field and not an outer imported accessor. #3801 would be a more complete fix for this and probably wouldn't break anything added here.
* The requested feature for automatically importing fixities from the record module doesn't seem practically achievable, as it would require the type of the record expression to be known during the concrete-to-abstract translation step, which is when fixities are relevant and AIUI strictly precedes type checking. An alternate syntax like `record MyRecord where`, where `MyRecord` is the name of the record module whence fixities can be read, might enable this feature; this PR doesn't preclude adding such a syntax in the future, though it's also the obvious niche for a not-verbose record update syntax in this style, and I'm currently agnostic as to which future feature seems more compelling.

### CODE REVIEW

File: CHANGELOG.md
Comment (jespercockx) [inline CHANGELOG.md:20]: Maybe you could add a minimal example here?

**Review by jespercockx — APPROVED** [review]
Thanks you for this PR, and my apologies that it took so long for it to get reviewed. Looking at the code, the implementation is remarkably uncomplicated which suggests that it's a robust way of leveraging existing infrastructure to extend the syntax in a practical way.

Given that we are very close to the release of 2.6.4 and I'd like this to get some proper testing before release, I propose we merge this PR right after the release (which should happen sometime in the next month). Could you please bring it back up-to-date with the current master by then?

File: CHANGELOG.md
Comment (rhendric) [reply CHANGELOG.md:20]: Done.

**Comment by andreasabel** [comment]
I rebased this onto current `master`, fixing the CHANGELOG conflict.

**Comment by andreasabel** [comment]
Thanks for this PR.  I agree that this is a reasonable sugar.

The general question is here whether the semantics should rather be defined in terms of _record-from-module_, e.g.
```agda
open import Agda.Builtin.Nat
open import Agda.Builtin.Sigma using (Σ)

test : Σ Nat λ _ → Nat
test = record{M} where
  module M where
    fst = 3
    snd = 5
```
The advantage of the latter is that one does not have the restriction of `let` in this interpretation, i.e., `M` can contain arbitrary declarations (in particular recursive ones and those that do pattern matching).
This desugaring does not work literally now if `fst` and/or `snd` are already in scope, e.g. imported from module `A.B.Sigma`.  This would not be a problem if we fixed a basic scoping issue first:
- #3801

**Comment by rhendric** [comment]
I remember considering that, and while I'm fuzzy on the details now, I think the main problem with it was that I wanted `record where` to be an expression and not just a declaration construct, and I didn't see a way for Agda's existing AST to support local modules in arbitrary expressions.

I probably could have worked around that by implementing some sort of module-lifting using parameterized modules, but it would have been extra work that might just have been ripped out in the long run depending on how/if #6149 comes to pass, and there might have been unanticipated semantic issues with that transformation, and as you note I would have been blocked on #3801 anyway, and all that would have been enough to persuade me to try the simpler thing first.

**Comment by jespercockx** [comment]
I think the proposal by Andreas is worth considering, but it is not a reason to postpone merging this PR, so I will go ahead and merge it now.

**Comment by rhendric** [comment]
Thank you!

**Comment by andreasabel** [comment]
@jespercockx wrote:
> I think the proposal by Andreas is worth considering, but it is not a reason to postpone merging this PR, so I will go ahead and merge it now.

In general, additions to the Agda syntax deserve a longer process making sure to reach a consensus.  After all, the concrete syntax is the thing that will stay in place even if you completely reimplement the type-checker/compiler.
This addition hasn't seen a discussion yet, I was just starting one.

E.g., the current semantics would make the concrete syntax `record let <ds>` a better fit as this communicates that the declarations are `let`-bindings and not `where`-bindings.

**Comment by jespercockx** [comment]
This PR has been open since May and I had announced well in advance my plan to merge it after the release, so I don't see why there hasn't been space for discussion.

**Comment by andreasabel** [comment]
When this PR was added in May my attention had already shifted to the release of 2.6.4 (scheduled for 1st July (!)) and I did not have the mental capacity to have a closer look at this language extension.  Apologies, @rhendric, that you had to wait longer here.

Anyway, I don't want to have this in without completing the discussion I just started.  There is no rush in merging a new feature, we will have 20 years to maintain it...

Let's continue at 
- #6911

**Comment by UlfNorell** [comment]
I prefer the `record where` syntax. `record let` looks strange to me: currently the `let` keyword is only used for let expressions, and this would add another quite different syntactic form using it. On the other hand `where` is already used to introduce blocks of various different things (where blocks to functions, data/record .. where, lambda where, do-binding where), so I don't expect having definitions in `record where` be restricted to let-declarations to cause too much confusion. Also, I believe we still have some ambitions to lift the restrictions on let-bindings at some point in the distant future.

**Comment by andreasabel** [comment]
Changing the milestone to `never` as this PR was moved to:
- #6911

**Pull request merged**
